File: ios/CruiseMesh/Core/FriendsOfFriendsStore.swift
import Foundation

enum FriendsOfFriendsStore {
    private static let enabledKey = "cruisemesh.friendsOfFriends.enabled"
    private static let revisionKey = "cruisemesh.friendsOfFriends.revision"
    private static let directoryRevisionKey = "cruisemesh.friendsOfFriends.directoryRevision"

    static func isEnabled() -> Bool {
        if AppDefaults.current.object(forKey: enabledKey) == nil { return true }
        return AppDefaults.current.bool(forKey: enabledKey)
    }

    static func revision() -> UInt64 {
        let existing = (AppDefaults.current.object(forKey: revisionKey) as? NSNumber)?.uint64Value ?? 0
        if existing > 0 { return existing }
        let initial = UInt64(max(1, Int64(Date().timeIntervalSince1970 * 1_000)))
        AppDefaults.current.set(NSNumber(value: initial), forKey: revisionKey)
        return initial
    }

    @discardableResult static func setEnabled(_ enabled: Bool) -> UInt64 {
        if isEnabled() == enabled { return revision() }
        let now = UInt64(max(1, Int64(Date().timeIntervalSince1970 * 1_000)))
        let next = max(revision() + 1, now)
        AppDefaults.current.set(enabled, forKey: enabledKey)
        AppDefaults.current.set(NSNumber(value: next), forKey: revisionKey)
        return next
    }

    static func nextDirectoryRevision() -> UInt64 {
        let existing = (AppDefaults.current.object(forKey: directoryRevisionKey) as? NSNumber)?.uint64Value ?? 0
        let now = UInt64(max(1, Int64(Date().timeIntervalSince1970 * 1_000)))
        let next = max(existing + 1, now)
        AppDefaults.current.set(NSNumber(value: next), forKey: directoryRevisionKey)
        return next
    }
}

File: android/app/src/main/kotlin/com/cruisemesh/app/friending/FriendsOfFriendsStore.kt
package com.cruisemesh.app.friending

import android.content.Context

/** Local policy advertised through profile-sync v2. */
object FriendsOfFriendsStore {
    private const val PREFS = "friends_of_friends"
    private const val KEY_ENABLED = "enabled"
    private const val KEY_REVISION = "revision"
    private const val KEY_DIRECTORY_REVISION = "directory_revision"

    fun isEnabled(context: Context): Boolean =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getBoolean(KEY_ENABLED, true)

    fun revision(context: Context): ULong {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val existing = prefs.getLong(KEY_REVISION, 0L)
        if (existing > 0L) return existing.toULong()
        val initial = System.currentTimeMillis().coerceAtLeast(1L)
        prefs.edit().putLong(KEY_REVISION, initial).apply()
        return initial.toULong()
    }

    fun setEnabled(context: Context, enabled: Boolean): ULong {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        if (prefs.getBoolean(KEY_ENABLED, true) == enabled) return revision(context)
        val next = maxOf(
            prefs.getLong(KEY_REVISION, 0L) + 1L,
            System.currentTimeMillis().coerceAtLeast(1L),
        )
        prefs.edit()
            .putBoolean(KEY_ENABLED, enabled)
            .putLong(KEY_REVISION, next)
            .apply()
        return next.toULong()
    }

    fun nextDirectoryRevision(context: Context): ULong {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val next = maxOf(
            prefs.getLong(KEY_DIRECTORY_REVISION, 0L) + 1L,
            System.currentTimeMillis().coerceAtLeast(1L),
        )
        prefs.edit().putLong(KEY_DIRECTORY_REVISION, next).apply()
        return next.toULong()
    }
}
